sequencer/src/l1_client.rs

```diff
@@ -195,15 +195,37 @@ async fn update_loop(
 
             // A new block has been produced. This happens fairly rarely, so it is now ok to poll to
             // see if a new block has been finalized.
-            let finalized = get_finalized_block(&rpc, finalized_tag, retry_delay).await;
+            let finalized = match get_finalized_block(&rpc, finalized_tag).await {
+                Ok(finalized) => Some(finalized),
+                Err(err) => {
+                    tracing::error!("failed to get finalized block: {err}");
+
+                    // If we cannot get the finalized block for any reason, don't let this stop us
+                    // from updating the L1 head. By returning `None` here, we will proceed to
+                    // update the head but not the finalized block, and then we will reset the
+                    // connection, in the hopes that this fixes whatever went wrong with
+                    // `get_finalized_block`.
+                    None
+                }
+            };
 
             // Update the snapshot.
             let mut snapshot = RwLockUpgradableReadGuard::upgrade(snapshot).await;
-            *snapshot = L1Snapshot { head, finalized };
+            snapshot.head = head;
+            if let Some(finalized) = finalized {
+                snapshot.finalized = finalized;
+            }
 
             // Drop our exclusive lock before logging anything.
             let snapshot = RwLockWriteGuard::downgrade(snapshot);
             tracing::info!("updated L1 snapshot to {:?}", *snapshot);
+
+            // If we encountered an error in `get_finalized_block`, reset the connection now that we
+            // have updated what we can.
+            if finalized.is_none() {
+                tracing::warn!("resetting connection due to error in get_finalized_block");
+                continue 'reset_connection;
```

Comment:
We could just break here - or is your idea to use the loop tag for a more semantic break?

Reply:
Idea was to avoid the redundant log after the inner loop but now that you mention it that seems like a bad reason to complicate the control flow. I'll change it to `break`